src/settings.rs

```diff
@@ -0,0 +1,140 @@
+use colors_transform::{Color, Rgb};
+use egui::Color32;
+use serde_derive::{Deserialize, Serialize};
+use std::{fs, io::Write, path::Path};
+
+#[derive(Deserialize, Serialize)]
+struct WasabiConfigFile {
+    note_speed: f64,
+    bg_color: String,
+    bar_color: String,
+    random_colors: bool,
+    sfz_path: String,
+    first_key: usize,
+    last_key: usize,
+}
+
+pub struct WasabiPermanentSettings {
+    pub note_speed: f64,
+    pub bg_color: Color32,
+    pub bar_color: Color32,
+    pub random_colors: bool,
+    pub sfz_path: String,
+    pub first_key: usize,
+    pub last_key: usize,
+}
+
+pub struct WasabiTemporarySettings {
+    pub panel_visible: bool,
+    pub stats_visible: bool,
+    pub settings_visible: bool,
+}
+
+impl Default for WasabiPermanentSettings {
+    fn default() -> Self {
+        WasabiPermanentSettings {
+            note_speed: 0.25,
+            bg_color: Color32::from_rgb(90, 90, 90),
+            bar_color: Color32::from_rgb(65, 0, 30),
+            random_colors: true,
+            sfz_path: "".to_string(),
+            first_key: 0,
+            last_key: 127,
+        }
+    }
+}
+
+static CONFIG_PATH: &str = "wasabi_config.toml";
+
+impl WasabiPermanentSettings {
+    pub fn new_or_load() -> Self {
+        if !Path::new(CONFIG_PATH).exists() {
+            let s = WasabiPermanentSettings::default();
+            let cfg = WasabiConfigFile {
+                note_speed: s.note_speed,
+                bg_color: Rgb::from(
+                    s.bg_color.r() as f32,
+                    s.bg_color.g() as f32,
+                    s.bg_color.b() as f32,
+                )
+                .to_css_hex_string(),
+                bar_color: Rgb::from(
+                    s.bar_color.r() as f32,
+                    s.bar_color.g() as f32,
+                    s.bar_color.b() as f32,
+                )
+                .to_css_hex_string(),
+                random_colors: s.random_colors,
+                sfz_path: s.sfz_path.clone(),
+                first_key: s.first_key,
+                last_key: s.last_key,
+            };
+            let toml: String = toml::to_string(&cfg).unwrap();
+            let mut file = fs::File::create(CONFIG_PATH).unwrap();
+            file.write_all(toml.as_bytes())
+                .expect("Error creating config");
+            s
+        } else {
+            let content = std::fs::read_to_string(CONFIG_PATH).unwrap();
+            let s: WasabiConfigFile = toml::from_str(&content).unwrap();
+            let bg_color_tmp = Rgb::from_hex_str(&s.bg_color).unwrap();
+            let bar_color_tmp = Rgb::from_hex_str(&s.bar_color).unwrap();
```

Comment:
For things like this, we need to eventually handle the edge cases properly. Every single .unwrap is a potential spot where the program could really crash.

We can leave it as-is for now as this is a nice start, but we do need to re-think the approach and the edge cases in the future.

Reply:
Will add a check function that validates that all values are correct before parsing anything.
Not the best solution, but I can't think of something better for now that will also keep the code clean.